crates/parser/src/parser.rs

```diff
@@ -386,4 +388,15 @@ impl CompletedMarker {
     pub(crate) fn kind(&self) -> SyntaxKind {
         self.kind
     }
+
+    pub(crate) fn last_token(&self, p: &Parser<'_>) -> Option<SyntaxKind> {
+        let end_pos = self.end_pos as usize;
+        if end_pos > p.events.len() {
+            return None;
+        }
+        p.events[..end_pos].iter().rev().find_map(|event| match event {
```

Comment:
I think this should be the following (to properly account for empty nodes)
```suggestion
        p.events[self.start_pos..end_pos].iter().rev().find_map(|event| match event {
```
BUT, the `forward_parent` stuff might cause issues for this and empty nodes aren't really a thing in r-a so keeping it as is ought to be fine